api/datastore/sql/sql.go

```diff
@@ -178,77 +125,6 @@ func pingWithRetry(ctx context.Context, attempts int, sleep time.Duration, db *s
 	return err
 }
 
-// checkExistence checks if tables have been created yet, it is not concerned
-// about the existence of the schema migration version (since migrations were
-// added to existing dbs, we need to know whether the db exists without migrations
-// or if it's brand new).
-func checkExistence(db *sqlx.DB) bool {
-	query := db.Rebind(`SELECT name FROM apps LIMIT 1`)
-	row := db.QueryRow(query)
-
-	var dummy string
-	err := row.Scan(&dummy)
-	if err != nil && err != sql.ErrNoRows {
-		// TODO we should probably ensure this is a certain 'no such table' error
-		// and if it's not that or err no rows, we should probably block start up.
-		// if we return false here spuriously, then migrations could be skipped,
-		// which would be bad.
-		return false
-	}
-	return true
-}
-
-// check if the db already existed, if the db is brand new then we can skip
-// over all the migrations BUT we must be sure to set the right migration
-// number so that only current migrations are skipped, not any future ones.
-func runMigrations(url string, exists bool) error {
-	m, err := migrator(url)
-	if err != nil {
-		return err
-	}
-	defer m.Close()
-
-	if !exists {
-		// set to highest and bail
-		return m.Force(latestVersion(migrations.AssetNames()))
-	}
-
-	// run any migrations needed to get to latest, if any
-	err = m.Up()
-	if err == migrate.ErrNoChange { // we don't care, but want other errors
-		err = nil
-	}
-	return err
-}
```

Comment:
i think we need to keep this logic so that fresh dbs skip all migrations still? this logic seems mostly drop in here.

Reply:
i disagree, even fresh datastore should be migrated since the tables were changed, so right after we create necessary tables we need to apply all migrations we have to make datastore consistent with Fn models.

Reply:
then we have to reconstruct the tables themselves from the deltas of all migrations in our head to know what a table looks like. cognitively, this is challenging with a non-trivial number of migrations.

Reply:
That's the point of the goose migrations, every change to the table (or a new table creation) should lay in migration, but not somewhere in the code. This is what we did a bit incorrectly at the very beginning. But this is what we have to do at this moment with goose, we **have** to reconstruct the tables by applying any number of the migrations we have to make datastore consistent.

Reply:
cat skinning...

if we kept schema files with the current state somewhere in our codebase it would be beneficial. as a dev, i don't want to go run sql queries against the db (nor is this plausible always) just to see what the db looks like before i go modify some fields in a patch.

Reply:
well, this is some kind of limitation that we have to live with, just as we lived with SQL only migrations, gonna have to get used to.

Reply:
we don't have to do anything, such is life :)

since we're going to have crippled CREATE TABLE to reference now can we find some way to have full schema of each table with the latest db and ensure in our tests that they actually work? right now we can reference `INSERT INTO` for a given table since fields are generally enumerated in that query, and this is tested in tests [generally], but this does not contain the type itself. one way to do this would be to in tests seed the db with the full schema to CREATE TABLE for each of our tables and then run all of our tests against the db, if this blows up then the schemas need to be updated (this would only happen in testing/CI). it's significantly less error prone and creates a significantly better user experience (when we have 100s of migrations to run on first boot) to just do this and skip migrations for fresh dbs...

Reply:
> it's significantly less error prone and creates a significantly better user experience (when we have 100s of migrations to run on first boot) to just do this and skip migrations for fresh dbs...

well, from my previous experience with other migrations frameworks (like one in Django, oslo.db, alembic), the number of migration to apply at the first boot guaranties that datastore 100% consistent, it happens once, so having migrations run once is totally fine, even in cluster, only the first Fn instance suppose to run all that migrations at once, any other instances will identify the migrations version and will run only the delta (from the particular version to the last one).

> since we're going to have crippled CREATE TABLE to reference now can we find some way to have full schema of each table with the latest db and ensure in our tests that they actually work?

This is what we had to worry about when we decided to use the migrations. At this moment we have no versioned objects or updated tests. So we don't know if the upgrade actually works until we test that manually. This is our fault that we have to fix.

Despite all of that, every time Fn will call datastore.New it will get the consistent datastore with latest migrations, so if the developer forgot to include the migration to the chain, it's his fault. This type of problems we already can catch with datastore tests as long as we test the migrations.

Reply:
> This type of problems we already can catch with datastore tests as long as we test the migrations.

slightly different problem: 

* me, developer, wants to change the type of a column
* how do i determine the type of that column? I must sift through all migrations or go ask a current db for the type of that column?

Reply:
> how do i determine the type of that column? I must sift through all migrations or go ask a current db for the type of that column?

Exactly, consistent datastore or the migrations are the sources of truth if you want to identify the column type.

Reply:
the last time i worked with a sql db was mostly rails, which active record very nicely handles maintaining a schema which is generated after migrations are run and this is dumped and checked into VCS and used to create new databases as well as for tests (what I'm hinting at). if we could automate that, great. maybe someone else can articulate what I mean: http://guides.rubyonrails.org/v3.2.8/migrations.html#schema-dumping-and-you -- we want/need an authoritative reference for each table, and there are some side benefits we can take advantage if we aren't religiously opposed to certain things.

Reply:
this kind of migrations works only for that particular framework, goose would fail every migration if the table would already contain exactly the same column, so it's bad behaviour in our particular case. So, having initial tables plus bunch of migrations (1-7) is definitely our path at this moment as long we use goose.

Reply:
I don't understand the concrete arguments for doing it this way other than that 'it's easier'. now that we at least try to skip the old migrations, I still do not understand the benefits of us running every migration on first boot and not maintaining a schema file where we have to figure out deltas as devs aside from the benefit of being able to skip the migrations. this seems to have nothing to do with the framework we're using, goose is very much just a library to do migrations, we can use it however we want. i see 0 blockers on using the same behavior as we had previously with migrate using goose, and 0 good reasons to switch to this newly proposed behavior. the old migrate library didn't maintain schema files either, we elected to keep an up to date schema in the CREATE TABLE blocks in our code and skip migrations for fresh boots. this new way proposed needs justification.